File: docs/ergonomics/scripts/matlab/cross_read.m
function cross_read(label, dir)
%CROSS_READ  Read another language's d1_family.mes, permute by name and
%   check the one deterministic value: instance 1, node 2, component 0
%   of the coordinates is 3.0.  Indices here are one based.

fprintf('##### d1_family written by %s #####\n', label);
d = mestra.read(fullfile(dir, 'd1_family.mes'));
fprintf('    %d rows\n', d.nRows);
fprintf('    keys: %s\n', strjoin(d.keyNames(), ', '));

c = d.support('s0').coordinates;
fprintf('    coordinates dims: %s\n', strjoin(c.dims, ', '));
p = mestra.permute(c.values, c.dims, {'group:member', 'node', 'component'});
fprintf('    permuted (group:member, node, component) size: %s\n', ...
        mat2str(size(p)));
fprintf('    value at instance 1, node 2, component 0 = %g\n', p(2, 3, 1));

e = d.nodeArray('s0', 'cad_edge_t');
fprintf('    cad_edge_t dims: %s\n', strjoin(e.dims, ', '));
q = mestra.permute(e.values, e.dims, {'node', 'component'});
fprintf('    cad_edge_t node 3 = %g\n', q(4, 1));

fprintf('    mach row 1 = %g\n', d.key('mach').values(2));
end

File: docs/ergonomics/scripts/julia/cross_read.jl
# Read another language's d1_family.mes, permute by name and check the
# one deterministic value.  Run: julia --project=julia cross_read.jl L=D ...

using Mestra

for arg in ARGS
    label, d = split(arg, "="; limit = 2)
    println("##### d1_family written by ", label, " #####")
    ds = Mestra.read(joinpath(d, "d1_family.mes"))
    println("    ", ds.nrows, " rows, aligned = ", ds.aligned)

    c = Mestra.values(ds, ds.supports[1].coordinates)
    println("    coordinates dims: ", Mestra.dimnames(c))
    p = Mestra.permute(c, (Symbol("group:member"), :node, :component))
    println("    permuted: ", Mestra.dimnames(p), " size ", size(p))
    println("    p[2, 3, 1] = ", p[2, 3, 1])
    println("    at(instance=2, node=3, component=1) = ",
            Mestra.at(c; instance = 2, node = 3, component = 1))

    e = Mestra.values(ds, ds["cad_edge_t"])
    println("    cad_edge_t dims: ", Mestra.dimnames(e),
            "  node 3 (0-based) = ", Mestra.at(e; node = 4, component = 1))
    # ds.keys["mach"].values is `nothing` after a lazy read and raises a
    # bare MethodError on Nothing.  Mestra.values is the documented way,
    # but the README never says the field exists and is empty.
    println("    mach row 1 (0-based) = ",
            Mestra.values(ds, ds.keys["mach"])[2])
end
